### PR TITLE
Oceanwater 534 implement current arm services as rosaction

### PR DESCRIPTION
**Pull request by anjan1984**
## Linked Issues:
| Jira Ticket 🎟️   | [Oceanwater-531]https://babelfish.arc.nasa.gov/jira/browse/OCEANWATER-531 |
| ----------- | ----------- |
| EPIC ⚡| [(OCEANWATER-518 | Switch to using effort_controllers/JointTrajectoryControllerl) |
| Github :octocat:  | # |

## Summary of Changes
* 
Implement available ros services as ros actions. All the services that have been converted to ros actions

1. stow 
2. unstow
3. Move guarded
3. grind
4. dig circular
5. dig linear
3. deliver sample 

## Test 
terminal 1:
roslaunch  ow europa_terminator_workspace.launch 

terminal 2:

- python unstow_action_client.py
- python stow_action_client.py
- python guarded_move_action_client.py
- python grind_action_client.py
- python dig_linear_action_client.py
- python dig_circular_action_client.py
- python deliver_action_client.py

look in client code for commented code to try default or custom values for each of the actions.

### CODE REVIEW

**Comment by kmdalal** [comment]
The tests ran fine.  Two small comments: the merge should be into melodic-devel and not master;  stow_action_client.py is missing.   I have more comments and questions to bring up in Dev Day.

**Review by Samahu — COMMENTED** [review]
ops, I had few comments pending for over two days. I forgot to publish .. sorry for the delay. However, I am not done yet.

File: ow_lander/action/Stow.action
Comment (kmdalal) [inline ow_lander/action/Stow.action:2]: What is the meaning of 'goal' ?

File: ow_lander/action/Stow.action
Comment (kmdalal) [inline ow_lander/action/Stow.action:12]: geometry_msgs/Point instead of x/y/z scalars is more clean and compact.

**Review by Samahu — CHANGES_REQUESTED** [review]
I am going to stop my review for this PR.

There are multiple issues with the current approach:
* Some action definitions are not clear
  - what is goal for unstow or stow
  - what is result
* Do we need to create an action server per activity?
  - Assuming we do (which I don't think it is the case), this will blow up very rapidly. I mean think of the amount of code that you have added for only three activities.
* For the other types of services there are more challenges lying ahead since the other service have dynamic parameters that can be set for every activity.
  - All the three activities that are implemented so far don't accept arguments that alter the arm trajectory at run time.

**Comment by kmdalal** [comment]
Results from testing so far (there are a few problems):

python unstow_action_client.py
 - good

python stow_action_client.py
 - good (did last)

python move_guarded_action.py
 - this file is missing

python grind_action_client.py
 - good

python dig_linear_action_client.py
 - scoop gets stuck in ground, LCP errors.   Perhaps because insufficient grind first?
 - Restarted sim, unstowed arm again.

python dig_circular_action_client.py
 - nothing happens (tried thrice, one non-consecutive)
 
python deliver_action_client.py
 - good

IN ALL CASES: in Rviz, the actual arm never moves, only the ghost arm.

**Comment by anjan1984** [comment]
Fixed linear trenching and circular dig. Should work now. I checked out melodic-devel and in RVIZ i see only the ghost arm move and not the actual arm

File: ow_lander/scripts/dig_circular_action_server.py
Comment (kmdalal) [inline ow_lander/scripts/dig_circular_action_server.py:42]: This line could be shortened for better readability.

File: ow_lander/scripts/dig_circular_action_server.py
Comment (kmdalal) [inline ow_lander/scripts/dig_circular_action_server.py:67]: Perhaps this issue is worth a Jira?

File: ow_lander/scripts/dig_linear_action_server.py
Comment (kmdalal) [inline ow_lander/scripts/dig_linear_action_server.py:38]: shorten line - stacked parameters would be much more readable

File: ow_lander/scripts/grind_action_server.py
Comment (kmdalal) [inline ow_lander/scripts/grind_action_server.py:36]: shorten line

File: ow_lander/scripts/stow_action_server.py
Comment (kmdalal) [inline ow_lander/scripts/stow_action_server.py:32]: shorten line

File: ow_lander/scripts/unstow_action_server.py
Comment (kmdalal) [inline ow_lander/scripts/unstow_action_server.py:32]: shorten line

**Comment by kmdalal** [comment]
> Fixed linear trenching and circular dig. Should work now. I checked out melodic-devel and in RVIZ i see only the ghost arm move and not the actual arm

Can all the actions work with any arguments now, not just the defaults?   Is only the ghost arm moving a problem?

**Comment by kmdalal** [comment]
I did not experience the TCP connection loss from client/action.  Could this problem be local to your environment?  Or is it a ROS problem to be expected in general?

**Comment by kmdalal** [comment]
Re-tried all the problematic action client tests I reported earlier, plus some others, and everything looked good!  

Ran ReferenceMission1 for regression (it uses Services) and it worked.

My code review was light (limited time, and I don't know Python well) but overall it looked clean and my comments were minor.

I'm done reviewing unless there are any other significant changes.

**Review by kmdalal — APPROVED** [review]

File: ow_lander/launch/planning.launch
Comment (Samahu) [reply ow_lander/launch/planning.launch:57]: You didn't respond to my inquiry, why did you need to define the node here

**Review by Samahu — CHANGES_REQUESTED** [review]
Based on  few testing I am seeing many issues when using values other than the default or when running the activity multiple time from a different state.
Here is some example: https://www.youtube.com/playlist?list=PLnVBqc6SfLfzxwClsf6XmuKfRUAAEQzes

File: ow_lander/action/Guardedmove.action
Comment (kmdalal) [inline ow_lander/action/Guardedmove.action:10]: Please rename this file (and action) to GuardedMove to be consistent with the others.

File: ow_lander/action/Guardedmove.action
Comment (kmdalal) [inline ow_lander/action/Guardedmove.action:6]: Anjan, please let me know when the ground-found success flag is added here.  Limited testing of the reference mission plans until then, as they rely on ground found before doing more arm ops.

Comment (kmdalal) [reply ow_lander/action/Guardedmove.action:6]: Actually, I'm faking ground found for now, so not blocked in testing the reference mission.

**Comment by kmdalal** [comment]
@anjan1984, only some of the Python scripts are directly executable.  For convenience and consistency, could you make them all (via chmod)?

**Comment by anjan1984** [comment]
will do the chmod changes

File: ow_lander/CMakeLists.txt
Comment (kmdalal) [inline ow_lander/CMakeLists.txt:37]: remove instead of comment-out

File: ow_lander/scripts/action_dig_circular.py
Comment (kmdalal) [inline ow_lander/scripts/action_dig_circular.py:35]: remove, or indicate why left commented

File: ow_lander/scripts/action_grind.py
Comment (kmdalal) [inline ow_lander/scripts/action_grind.py:95]: remove if no longer useful

**Comment by kmdalal** [comment]
I'm done with my second review.  Went light on the code due to limited time and knowledge of Python - mainly looked for readability.   I plan to test the actions more from autonomy, and issues I find could be addressed by new Jiras.

**Comment by Samahu** [comment]
@anjan1984 I started testing this PR but the The PR description still says it is a work of progress

**Comment by anjan1984** [comment]
> @anjan1984 I started testing this PR but the The PR description still says it is a work of progress

updated the description

**Review by Samahu — CHANGES_REQUESTED** [review]
Sorry @anjan1984 but this is still not working for me :disappointed: 
https://youtu.be/PHD72emMGG8
https://youtu.be/uI6FOPr5aPM

@kmdalal @mogumbo Can these weird artifacts be ignored? I really don't wish to be in the position in which I keep breaking this every time.

**Comment by kmdalal** [comment]
> Sorry @anjan1984 but this is still not working for me 
> https://youtu.be/PHD72emMGG8
> https://youtu.be/uI6FOPr5aPM
> 
> @kmdalal @mogumbo Can these weird artifacts be ignored? I really don't wish to be in the position in which I keep breaking this every time.

I have seen this weird dig_circular movement many times.  I had not seen this GuardedMove issue.  I think they need to be fixed -- the question is whether to push for that in our next release or leave them as caveats and create new Jiras for them...

**Comment by anjan1984** [comment]
> Sorry @anjan1984 but this is still not working for me 
> https://youtu.be/PHD72emMGG8
> https://youtu.be/uI6FOPr5aPM
> 
I am working on these issues to see if I can fix them

**Comment by anjan1984** [comment]
> > > Sorry @anjan1984 but this is still not working for me
> > > https://youtu.be/PHD72emMGG8
> > > https://youtu.be/uI6FOPr5aPM
> > > @kmdalal @mogumbo Can these weird artifacts be ignored? I really don't wish to be in the position in which I keep breaking this every time.
> > 
> > 
> > I have seen this weird dig_circular movement many times. I had not seen this GuardedMove issue. I think they need to be fixed -- the question is whether to push for that in release 8 or leave them as caveats and create new Jiras for them...
> 
> You probably mean the grind movement .. I didn't attempt to test GuardedMove.

Latest commits should fix the problems

File: ow_lander/scripts/action_deliver_sample.py
Comment (mogumbo) [inline ow_lander/scripts/action_deliver_sample.py:52]: Different tab sizes in the same file. Please fix.

**Review by mogumbo — CHANGES_REQUESTED** [review]
The only runtime problem I spotted was the scoop strays pretty far off its target when it starts dumping during sample delivery.

**Comment by Samahu** [comment]
> The only runtime problem I spotted was the scoop strays pretty far off its target when it starts dumping during sample delivery.

I agree with Terry, there is a noticeable change in the scoop position from what we currently have.

**Comment by Samahu** [comment]
I have noticed that "Guarded Move" halts movement without coming into contact with the ground:

This is the most common:
image

Sometimes too early:
image

Does this branch rely on the F/T sensor when checking for contact with the ground?

**Comment by Samahu** [comment]
image

**Comment by anjan1984** [comment]
> The only runtime problem I spotted was the scoop strays pretty far off its target when it starts dumping during sample delivery.

I get this movement 

https://www.youtube.com/watch?v=aQGWkxT8tcM

Created Jira issue to further investigate https://babelfish.arc.nasa.gov/jira/browse/OCEANWATER-661

**Comment by Samahu** [comment]
I am observing an issues when trying to repeat the guarded move .. this one of the things that we spent time in past releases to make sure the operation can be performed multiple times in the same session.

**Comment by anjan1984** [comment]
> I have noticed that "Guarded Move" halts movement without coming into contact with the ground:
> 
> This is the most common:
> image
> 
> Sometimes even too early:
> image
> 
> Does this branch rely on the F/T sensor when checking for contact with the ground?

No . Created new ticket to change to F/T sensor for checking the ground 

https://babelfish.arc.nasa.gov/jira/browse/OCEANWATER-662

**Comment by Samahu** [comment]
> > I have noticed that "Guarded Move" halts movement without coming into contact with the ground:
> > Does this branch rely on the F/T sensor when checking for contact with the ground?
> 
> No . Created new ticket to change to F/T sensor for checking the ground
> 
> https://babelfish.arc.nasa.gov/jira/browse/OCEANWATER-662

Checking for ground detection has already been merged to melodic-devel .. so hopefully if sync this branch with it things will look better .. I hope.

**Comment by anjan1984** [comment]
> > > I have noticed that "Guarded Move" halts movement without coming into contact with the ground:
> > > Does this branch rely on the F/T sensor when checking for contact with the ground?
> > 
> > 
> > No . Created new ticket to change to F/T sensor for checking the ground
> > https://babelfish.arc.nasa.gov/jira/browse/OCEANWATER-662
> 
> Checking for ground detection using the F/T sensor has already been merged to melodic-devel .. so hopefully if sync this branch with it things will look better .. I hope.

I think so

**Comment by mogumbo** [comment]
> The only runtime problem I spotted was the scoop strays pretty far off its target when it starts dumping during sample delivery.

I just updated to 005d310d6efa05615bce578ffd0e11e322b7a45b and the behavior has changed. If I start the sim and immediately do `python deliver_action_client.py` the scoop moves over the delivery target and stops. No dumping occurs.

**Comment by kmdalal** [comment]
I'm resuming my review.  GuardedMove fails to find ground as executed from ReferenceMission1.  The following error is produced:

```
[ERROR]: bad callback: <bound method ActionClient._result_cb of <actionlib.action_client.ActionClient instance at 0x7f68a1850140>>                                            
Traceback (most recent call last):                                                     
  File "/opt/ros/melodic/lib/python2.7/dist-packages/rospy/topics.py", line 750, in _invoke_callback                                                                          
    cb(msg)                                                                            
  File "/opt/ros/melodic/lib/python2.7/dist-packages/actionlib/action_client.py", line 625, in _result_cb                                                                     
    self.manager.update_results(msg)                                                   
  File "/opt/ros/melodic/lib/python2.7/dist-packages/actionlib/action_client.py", line 492, in update_results                                                                 
    status.update_result(action_result)                                                
  File "/opt/ros/melodic/lib/python2.7/dist-packages/actionlib/action_client.py", line 404, in update_result                                                                  
    self.transition_to(CommState.DONE)                                                 
  File "/opt/ros/melodic/lib/python2.7/dist-packages/actionlib/action_client.py", line 377, in transition_to                                                                  
    self.transition_cb(ClientGoalHandle(self))                                         
  File "/opt/ros/melodic/lib/python2.7/dist-packages/actionlib/simple_action_client.py", line 241, in _handle_transition                                                      
    self.done_cb(gh.get_goal_status(), gh.get_result())                                
  File "/usr/local/home/kdalal/ow/src/ow_simulator/ow_lander/scripts/guarded_move_action_server.py", line 50, in handle_guarded_move_done                                     
    ground_position = self.ground_detector.ground_position if ground_detected else Point()                                                                                    
NameError: global name 'Point' is not defined                                          

[ERROR]: Exception in your execute callback: GroundDetector instance has no attribute '_last_position'                                                                        
Traceback (most recent call last):                                                     
  File "/opt/ros/melodic/lib/python2.7/dist-packages/actionlib/simple_action_server.py", line 289, in executeLoop                                                             
    self.execute_callback(goal)                                                        
  File "/usr/local/home/kdalal/ow/src/ow_simulator/ow_lander/scripts/guarded_move_action_server.py", line 116, in on_guarded_move_action                                      
    self._result.final.x = self.ground_detector._last_position[0]                      
AttributeError: GroundDetector instance has no attribute '_last_position'    
```

**Comment by kmdalal** [comment]
@anjan1984 , could you please reduce the INFO messages, e.g. turn them into DEBUG?  They are quite verbose.

File: ow_lander/scripts/LanderInterface.py
Comment (kmdalal) [inline ow_lander/scripts/LanderInterface.py:86]: Remove, or add comment if left in on purpose.

File: ow_lander/scripts/action_deliver_sample.py
Comment (kmdalal) [inline ow_lander/scripts/action_deliver_sample.py:16]: Remove if not needed.

**Review by kmdalal — CHANGES_REQUESTED** [review]
ReferenceMission1 and TestActions work fine and look good.  Just requested minor code changes -- all formatting.

**Comment by anjan1984** [comment]
> > The only runtime problem I spotted was the scoop strays pretty far off its target when it starts dumping during sample delivery.
> 
> I just updated to [005d310](https://github.com/nasa/ow_simulator/commit/005d310d6efa05615bce578ffd0e11e322b7a45b) and the behavior has changed. If I start the sim and immediately do `python deliver_action_client.py` the scoop moves over the delivery target and stops. No dumping occurs.

New updates should solve this issue

**Comment by anjan1984** [comment]
> ReferenceMission1 and TestActions work fine and look good. Just requested minor code changes -- all formatting.

Thank you. Made all the changes you requested

**Comment by anjan1984** [comment]
> > > I have noticed that "Guarded Move" halts movement without coming into contact with the ground:
> > > Does this branch rely on the F/T sensor when checking for contact with the ground?
> > 
> > 
> > No . Created new ticket to change to F/T sensor for checking the ground
> > https://babelfish.arc.nasa.gov/jira/browse/OCEANWATER-662
> 
> Checking for ground detection using the F/T sensor has already been merged to melodic-devel .. so hopefully if sync this branch with it things will look better .. I hope.

latest commits should fix this issue

**Review by kmdalal — APPROVED** [review]

**Comment by Samahu** [comment]
> > > > I have noticed that "Guarded Move" halts movement without coming into contact with the ground:
> > > > Does this branch rely on the F/T sensor when checking for contact with the ground?
> > > 
> > > 
> > > No . Created new ticket to change to F/T sensor for checking the ground
> > > https://babelfish.arc.nasa.gov/jira/browse/OCEANWATER-662
> > 
> > 
> > Checking for ground detection using the F/T sensor has already been merged to melodic-devel .. so hopefully if sync this branch with it things will look better .. I hope.
> 
> latest commits should fix this issue

The changes you made seem to fix premature stop that I shown in the images but it still doesn't address that Guarded Move is no longer repeatable. I can easily produce the issue using the current defaults, simply run: **unstow**, **guarded_move**, then another **guarded_move**.

**Comment by anjan1984** [comment]
> > > > > I have noticed that "Guarded Move" halts movement without coming into contact with the ground:
> > > > > Does this branch rely on the F/T sensor when checking for contact with the ground?
> > > > 
> > > > 
> > > > No . Created new ticket to change to F/T sensor for checking the ground
> > > > https://babelfish.arc.nasa.gov/jira/browse/OCEANWATER-662
> > > 
> > > 
> > > Checking for ground detection using the F/T sensor has already been merged to melodic-devel .. so hopefully if sync this branch with it things will look better .. I hope.
> > 
> > 
> > latest commits should fix this issue
> 
> The changes you made seem to fix premature stop that I shown in the images but it still doesn't address that Guarded Move is no longer repeatable. I can easily produce the issue using the current defaults, simply run: **unstow**, **guarded_move**, then another **guarded_move**. When you do that this may or may not result in a movement for the second attempt.

Fixed!

**Comment by Samahu** [comment]
I did another test and got same results .. I’ll differ this PR completely to @mikedalal and @mogumbo.

**Comment by kmdalal** [comment]
I committed executable versions of the action client scripts, for convenience, and to be consistent with the other scripts.

**Comment by kmdalal** [comment]
I tried @Samahu's double GuardedMove test using Atacama.  The first one succeeded and looked good.  The second succeeded and there was very little movement, but that's what I expected since the scoop was already touching the ground.  Is this the correct behavior?  

I retested most of the other actions, using the scripts and the TestActions plan (default versions), and overall it looked fine.  I did notice some odd movements here and there, but I think we should address separately after the release.  Some of these path issues may have existed in the Services.

**Comment by anjan1984** [comment]
> I tried @Samahu's double GuardedMove test using Atacama. The first one succeeded and looked good. The second succeeded and there was very little movement, but that's what I expected since the scoop was already touching the ground. Is this the correct behavior?
> 

This is not the expected behavior. The expected behavior is repeating the same movement and finding the ground again. But the solution is complicated. The way the system is designed is the ground is found when the FT sensor reads a value greater that a threshold. In services this is implemented only on the last segment of the trajectory. In actions, however, we concatenate all the trajectories and send it to the trajectory executer. During the entire trajectory the FT sensors reads that threshold many times and stops in between. To overcome this problem, I reset the sensor if the position of the scoop is not close to the ground. Hence it finds the ground perfectly. Now If you are repeating the GuardedMove from a GuardedMove position the FT sensor reads a reading during the initial jerk and since it is close to the ground it is not reset. 

Instead of double Guardedmove do Guardedmove Unstow and then Guardedmove again and you would see the expected behavior. 

As I mentioned it has a complicated solution and will create a separate ticket to handle this (if needed).?

**Comment by kmdalal** [comment]
> As I mentioned it has a complicated solution and will create a separate ticket to handle this (if needed).?

Thanks @anjan1984 for the detailed explanation.  I suggest putting all this into a new Jira and dealing with it after the release.  And to document the workaround (unstow in between guarded moves) in the user guide.   Successive guarded moves could actually be part of a workspace  interrogation scheme.

**Comment by anjan1984** [comment]
> > As I mentioned it has a complicated solution and will create a separate ticket to handle this (if needed).?
> 
> Thanks @anjan1984 for the detailed explanation. I suggest putting all this into a new Jira and dealing with it after the release. And to document the workaround (unstow in between guarded moves) in the user guide. Successive guarded moves could actually be part of a workspace interrogation scheme.


Created this ticket 
https://babelfish.arc.nasa.gov/jira/browse/OCEANWATER-668

**Review by Samahu — APPROVED** [review]
I did test this PR several times and things has improved overall. Yet, the ROS action version of certain arm activities has different behavior than their ROS services counterparts. I personally wouldn't expect this to be the case but I don't want to block the merge if others see this as insignificant and can be addressed later. Thus, I am differing the final decision to the other reviewers as I am done testing the PR.

I won't be doing a code review this time due to the limited time and my [original objection on the approach taken for creating a ros action per each individual activity](https://github.com/nasa/ow_simulator/pull/66#pullrequestreview-573722174).

**Review by kmdalal — APPROVED** [review]

**Review by mogumbo — APPROVED** [review]
Approved. I still see the scoop dumping in the wrong place when running deliver_action_client.py. This is only a cosmetic problem, so please make a Jira ticket to fix it but it's low priority.

File: ow_lander/scripts/guarded_move_action_server.py
Comment (kmdalal) [inline ow_lander/scripts/guarded_move_action_server.py:124]: Please rename to 'GuardedMove' so that we have consistent naming (CamelCase) for all actions.

**Review by kmdalal — CHANGES_REQUESTED** [review]
I just requested a simple rename in the guarded move action server/client, and I would need to make the same change in my autonomy branch.

File: ow_lander/scripts/guarded_move_action_client.py
Comment (kmdalal) [inline ow_lander/scripts/guarded_move_action_client.py:15]: Please rename to 'GuardedMove' to be consistent with the other actions.

**Comment by lanseafood** [comment]
@anjan1984 @mikedalal @Samahu Was playing around to start the rest of the arm faults but I notice that in this PR the version of arm faults for Unstow and Stow are broken. Faults in the arm no longer cause a stop in the arm movement in Gazebo.

**Pull request merged**
